ESPS/general/man/man3/alaw.3
.\" Copyright (c) 1986-1990 Entropic Speech, Inc.
.\" Copyright (c) 1991 Entropic Research Laboratory, Inc. All rights reserved.
.\" @(#)alaw.3	1.5 24 Jul 1991 ESI/ERL
.ds ]W (c) 1991 Entropic Research Laboratory, Inc.
.TH ALAW 3\-ESPSsp 24 Jul 1991
.SH "NAME"

.nf
linear_to_a \- A-law compression of linear PCM data 
a_to_linear \- re-expansion of A-law compressed data
linear_to_a_2 \- A-law compression (with bit inversions) of linear PCM data 
a_to_linear_2 \- re-expansion of A-law compressed data (with bit inversions)
.fi
.SH "SYNOPSIS"
.nf
.ft B
int
linear_to_a(inbuf, outbuf, bufsize)
short	*inbuf;
char	*outbuf;
long	bufsize;

int
a_to_linear(inbuf, outbuf, bufsize)
char	*inbuf;
short	*outbuf;
long	bufsize;

int
linear_to_a_2(inbuf, outbuf, bufsize)
short	*inbuf;
char	*outbuf;
long	bufsize;

int
a_to_linear_2(inbuf, outbuf, bufsize)
char	*inbuf;
short	*outbuf;
long	bufsize;
.ft
.fi
.SH "DESCRIPTION"
.PP
These functions do quasi-logarithmic companding (compression and expansion)
of 2-byte integer data by using the
.IR A -law
with a value of
.I A
= 87.56.
This is the European PCM standard [1].
This compression law is linear for small amplitudes
and logarithmic for large amplitudes.
The function takes input data to be converted from
.I inbuf
and outputs the converted data to
.I outbuf;
the argument
.I bufsize
is the number of elements in the input array.
.PP
.I linear_to_a
converts integer data with 13 significant bits (including sign) to 8 bits
by using a 13-segment, piecewise linear approximation to the
.IR A -law
compression characteristic [1].
An input value greater than 4095 or less than \-4096
gives the same output code as 4095 or \-4096.
.PP
.I a_to_linear
converts 8-bit, logarithmically compressed data to 13-bit linear data
according to the piecewise-linear
.IR A -law
approximation.
A table lookup procedure is used to decode the the compressed values.
.PP
.I linear_to_a_2
and
.I a_to_linear_2
are versions of
.I linear_to_a
and
.I a_to_linear
that take into account
the CCITT recommended practice of inverting certain bits of the
.IR A -law
encoded data before transmission [2].
(This is done to even out the density of
.IR 1 s
and
.IR 0s
in the output bit stream for low-level inputs.)
It is easy to transform the output of
.I linear_to_a
to that of
.I linear_to_a_2,
or vice versa:  modify each output byte by taking the exclusive
.I or
with the constant
.I 0xd5
(binary 11010101).
Likewise,
.I a_to_linear_2
gives the same output that 
.I a_to_linear
would give if each input byte were modified in the same way.
Companding by applying
.I a_to_linear
to the output of
.I linear_to_a
gives exactly the same result as using
.I a_to_linear_2
and
.I linear_to_a_2
instead.
The latter two functions are intended for use
when the coded form must be compatible with existing telecommunications data.
If a tractable numerical relation between the input and output data
is more important, then
.I a_to_linear
and
.I linear_to_a
are the better choice.
.SH EXAMPLES
.PP
Running
.RS
.nf

int i;
static short original[13] =
	{\-4096, \-5, \-4, \-3, \-2, \-1, 0, 1, 2, 3, 4, 5, 4095};
char encoded[13];
short decoded[13];

linear_to_a(original, encoded, (long) 13);
a_to_linear(encoded, decoded, (long) 13);

for (i = 0; i < 13; i++) printf("%5d", original[i]);
printf("\\n");
for (i = 0; i < 13; i++) printf("%5x", encoded[i]&0xff);
printf("\\n");
for (i = 0; i < 13; i++) printf("%5d", decoded[i]);
printf("\\n");

.fi
.RE
prints the output
.nf
.if n .ta .5iR +5nR +5nR +5nR +5nR +5nR +5nR +5nR +5nR +5nR +5nR +5nR +5nR +5nR
.if t .ta .5iR +6nR +6nR +6nR +6nR +6nR +6nR +6nR +6nR +6nR +6nR +6nR +6nR +6nR

	\-4096	\-5	\-4	\-3	\-2	\-1	0	1	2	3	4	5	4095	
	ff	82	81	81	80	80	0	0	1	1	2	2	7f	
	\-4032	\-5	\-3	\-3	\-1	\-1	1	1	3	3	5	5	4032	

.fi
.DT
Running the same code with the function names changed to
.I linear_to_a_2
and
.I a_to_linear_2
prints the output
.if n .ta .5iR +5nR +5nR +5nR +5nR +5nR +5nR +5nR +5nR +5nR +5nR +5nR +5nR +5nR
.if t .ta .5iR +6nR +6nR +6nR +6nR +6nR +6nR +6nR +6nR +6nR +6nR +6nR +6nR +6nR
.nf

	\-4096	\-5	\-4	\-3	\-2	\-1	0	1	2	3	4	5	4095	
	2a	57	54	54	55	55	d5	d5	d4	d4	d7	d7	aa	
	\-4032	\-5	\-3	\-3	\-1	\-1	1	1	3	3	5	5	4032	

.fi
.DT
.SH "BUGS"
.PP
None Known.
.SH REFERENCES
.PP
[1] John C. Bellamy,
.I Digital Telephony,
(New York: Wiley, 1982), Appendix B.
.PP
[2] International Telegraph and Telephone Consultative Committee (CCITT)
.I Orange Book
v. III-2 (1976), Recommendation G.711.
.SH "SEE ALSO"
.PP
\fIlinear_to_mu\fP (3\-\s-1ESPS\s+1), 
\fImu_to_linear\fP (3\-\s-1ESPS\s+1)
.SH "AUTHOR"
.PP
Manual page and programs by David Burton.
Modified by Rodney Johnson.


ESPS/general/src/lib_sp/alaw.c
/*
 * This material contains unpublished, proprietary software of 
 * Entropic Research Laboratory, Inc. Any reproduction, distribution, 
 * or publication of this work must be authorized in writing by Entropic 
 * Research Laboratory, Inc., and must bear the notice: 
 *
 *    "Copyright (c) 1986-1990  Entropic Speech, Inc. 
 *    "Copyright (c) 1990-1991  Entropic Research Laboratory, Inc. 
 *                   All rights reserved"
 *
 * The copyright notice above does not evidence any actual or intended 
 * publication of this source code.     
 *
 * Written by:  David Burton
 * Checked by:
 * Revised by:  Rodney Johnson
 *
 * Functions linear_to_a, a_to_linear, linear_to_a_2, a_to_linear_2
 * convert linear (PCM coded) data to A-law data and back.
 * The "_2" functions invert code bits according to CCITT convention.
 */

static char *sccs_id = "@(#)alaw.c	1.2	7/24/91	ESI/ERL";

/*
 * TABLES FOR CONVERSIONS
 */

static short A_seg_table[128] = {
 0, 1, 2, 2, 3, 3, 3, 3, 4, 4, 4, 4, 4, 4, 4, 4,
 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 
 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 
 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 
 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 
 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 
 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 
 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 
};

/**//* Decoding table for a_to_linear created by the following. */
/* #include <stdio.h>							*/
/* static short    A_dec_table[256];					*/
/* main() {								*/
/*     int i;								*/
/*     for (i = 0; i < 32; i++) A_dec_table[i] = 2*i + 1;		*/
/*     for ( ; i < 128; i++)    A_dec_table[i] = 2*A_dec_table[i-16];	*/
/*     for ( ; i < 256; i++)    A_dec_table[i] = -A_dec_table[i-128];	*/
/*     for (i = 0; i < 256; i++) {					*/
/*         printf("%6d,", A_dec_table[i]);				*/
/*         if ((i+1)%8 == 0) printf("\n");				*/
/*     }								*/
/*     exit(0);								*/
/* }									*/
/**/

/* Decoding table for a_to_linear. */

static short A_dec_table[256] = {
     1,     3,     5,     7,     9,    11,    13,    15,
    17,    19,    21,    23,    25,    27,    29,    31,
    33,    35,    37,    39,    41,    43,    45,    47,
    49,    51,    53,    55,    57,    59,    61,    63,
    66,    70,    74,    78,    82,    86,    90,    94,
    98,   102,   106,   110,   114,   118,   122,   126,
   132,   140,   148,   156,   164,   172,   180,   188,
   196,   204,   212,   220,   228,   236,   244,   252,
   264,   280,   296,   312,   328,   344,   360,   376,
   392,   408,   424,   440,   456,   472,   488,   504,
   528,   560,   592,   624,   656,   688,   720,   752,
   784,   816,   848,   880,   912,   944,   976,  1008,
  1056,  1120,  1184,  1248,  1312,  1376,  1440,  1504,
  1568,  1632,  1696,  1760,  1824,  1888,  1952,  2016,
  2112,  2240,  2368,  2496,  2624,  2752,  2880,  3008,
  3136,  3264,  3392,  3520,  3648,  3776,  3904,  4032,
    -1,    -3,    -5,    -7,    -9,   -11,   -13,   -15,
   -17,   -19,   -21,   -23,   -25,   -27,   -29,   -31,
   -33,   -35,   -37,   -39,   -41,   -43,   -45,   -47,
   -49,   -51,   -53,   -55,   -57,   -59,   -61,   -63,
   -66,   -70,   -74,   -78,   -82,   -86,   -90,   -94,
   -98,  -102,  -106,  -110,  -114,  -118,  -122,  -126,
  -132,  -140,  -148,  -156,  -164,  -172,  -180,  -188,
  -196,  -204,  -212,  -220,  -228,  -236,  -244,  -252,
  -264,  -280,  -296,  -312,  -328,  -344,  -360,  -376,
  -392,  -408,  -424,  -440,  -456,  -472,  -488,  -504,
  -528,  -560,  -592,  -624,  -656,  -688,  -720,  -752,
  -784,  -816,  -848,  -880,  -912,  -944,  -976, -1008,
 -1056, -1120, -1184, -1248, -1312, -1376, -1440, -1504,
 -1568, -1632, -1696, -1760, -1824, -1888, -1952, -2016,
 -2112, -2240, -2368, -2496, -2624, -2752, -2880, -3008,
 -3136, -3264, -3392, -3520, -3648, -3776, -3904, -4032,
};

/**//* Decoding table for a_to_linear_2 created by the following. */
/* #include <stdio.h>							*/
/* static short    A_dec_tbl_2[256];					*/
/* main() {								*/
/*     int i;								*/
/*     for (i = 0; i < 32; i++) A_dec_tbl_2[i^0x55] = -(2*i + 1);	*/
/*     for ( ; i < 128; i++)    A_dec_tbl_2[i^0x55] =			*/
/* 					    2*A_dec_tbl_2[(i-16)^0x55]; */
/*     for ( ; i < 256; i++)    A_dec_tbl_2[i] = -A_dec_tbl_2[i-128];	*/
/*     for (i = 0; i < 256; i++) {					*/
/*         printf("%6d,", A_dec_tbl_2[i]);				*/
/*         if ((i+1)%8 == 0) printf("\n");				*/
/*     }								*/
/*     exit(0);								*/
/* }									*/
/**/

/* Decoding table for a_to_linear_2. */

static short A_dec_tbl_2[256] = {
  -688,  -656,  -752,  -720,  -560,  -528,  -624,  -592,
  -944,  -912, -1008,  -976,  -816,  -784,  -880,  -848,
  -344,  -328,  -376,  -360,  -280,  -264,  -312,  -296,
  -472,  -456,  -504,  -488,  -408,  -392,  -440,  -424,
 -2752, -2624, -3008, -2880, -2240, -2112, -2496, -2368,
 -3776, -3648, -4032, -3904, -3264, -3136, -3520, -3392,
 -1376, -1312, -1504, -1440, -1120, -1056, -1248, -1184,
 -1888, -1824, -2016, -1952, -1632, -1568, -1760, -1696,
   -43,   -41,   -47,   -45,   -35,   -33,   -39,   -37,
   -59,   -57,   -63,   -61,   -51,   -49,   -55,   -53,
   -11,    -9,   -15,   -13,    -3,    -1,    -7,    -5,
   -27,   -25,   -31,   -29,   -19,   -17,   -23,   -21,
  -172,  -164,  -188,  -180,  -140,  -132,  -156,  -148,
  -236,  -228,  -252,  -244,  -204,  -196,  -220,  -212,
   -86,   -82,   -94,   -90,   -70,   -66,   -78,   -74,
  -118,  -114,  -126,  -122,  -102,   -98,  -110,  -106,
   688,   656,   752,   720,   560,   528,   624,   592,
   944,   912,  1008,   976,   816,   784,   880,   848,
   344,   328,   376,   360,   280,   264,   312,   296,
   472,   456,   504,   488,   408,   392,   440,   424,
  2752,  2624,  3008,  2880,  2240,  2112,  2496,  2368,
  3776,  3648,  4032,  3904,  3264,  3136,  3520,  3392,
  1376,  1312,  1504,  1440,  1120,  1056,  1248,  1184,
  1888,  1824,  2016,  1952,  1632,  1568,  1760,  1696,
    43,    41,    47,    45,    35,    33,    39,    37,
    59,    57,    63,    61,    51,    49,    55,    53,
    11,     9,    15,    13,     3,     1,     7,     5,
    27,    25,    31,    29,    19,    17,    23,    21,
   172,   164,   188,   180,   140,   132,   156,   148,
   236,   228,   252,   244,   204,   196,   220,   212,
    86,    82,    94,    90,    70,    66,    78,    74,
   118,   114,   126,   122,   102,    98,   110,   106,
};


/*
 * FUNCTION DEFINITIONS
 */

/* Functions to convert to A-encoded data */

int
linear_to_a(input, output, size)	/* non-CCITT */
    short	*input;		/* PCM input */
    char	*output;	/* A-encoded output */
    long	size;		/* number of points in input */
{
    long	i;		/* loop counter */
    unsigned	in;		/* holds input magnitude */
    unsigned    P;		/* sign bit: 0x80 if negative, else 0x00 */
    unsigned	S;		/* segment identifier: range 0 - 7 */
    unsigned	Q;		/* quantized step: range 0 - 15 */

    for(i = 0; i < size; i++)
    {
	if (input[i] < 0)
	{
	    P = 0x80;
	    in = -1 - input[i];
	}
	else
	{
	    P = 0x00;
	    in = input[i];
	}

	if (in > 4095) in = 4095;

	S = A_seg_table[in >> 5];

	if (S == 0)
	    Q = (in >> 1) & 0x0f;
	else
	    Q = (in >> S) & 0x0f;

	output[i] = P | (S << 4) | Q;
    }

    return 0;
}

int
linear_to_a_2(input, output, size)	/* CCITT */
    short	*input;		/* PCM input */
    char	*output;	/* A-encoded output */
    long	size;		/* number of points in input */
{
    long	i;		/* loop counter */
    unsigned	in;		/* holds input magnitude */
    unsigned    P;		/* sign bit: 0x00 if negative, else 0x80 */
    unsigned	S;		/* segment identifier: range 0 - 7 */
    unsigned	Q;		/* quantized step: range 0 - 15 */

    for(i = 0; i < size; i++)
    {
	if (input[i] < 0)
	{
	    P = 0x00;		/* Opposite sign from linear_to_a. */
	    in = -1 - input[i];
	}
	else
	{
	    P = 0x80;		/* Opposite sign from linear_to_a. */
	    in = input[i];
	}

	if (in > 4095) in = 4095;

	S = A_seg_table[in >> 5];

	if (S == 0)
	    Q = (in >> 1) & 0x0f;
	else
	    Q = (in >> S) & 0x0f;
				/* Alternate magnitude bits inverted. */
	output[i] = (P | (S << 4) | Q) ^ 0x55;
    }

    return 0;
}

/* Functions to decode A-coded values */

int
a_to_linear(in, out, size)		/* non-CCITT */
    char	*in;		/*input A-coded values*/
    short	*out;		/*output linear values*/
    long	size;		/*number of items in the input array*/
{
    long	i;		/*loop counter*/

    for(i = 0; i < size; i++)
	out[i] = A_dec_table[(unsigned char) in[i]];

    return 0;
}


int
a_to_linear_2(in, out, size)		/* CCITT */
    char	*in;		/*input A-coded values*/
    short	*out;		/*output linear values*/
    long	size;		/*number of items in the input array*/
{
    long	i;		/*loop counter*/

    for(i = 0; i < size; i++)
	out[i] = A_dec_tbl_2[(unsigned char) in[i]];

    return 0;
}

